src/prof_recent.c

```diff
@@ -0,0 +1,428 @@
+	assert(recent_alloc == extent_prof_recent_alloc_get(tsd, extent));
+	extent->e_prof_info.e_prof_recent_alloc = NULL;
+	recent_alloc->alloc_extent = NULL;
+}
+
+/*
+ * This function should be called right before an allocation is released, so
+ * that the associated recent allocation record can contain the following
+ * information:
+ * (1) The allocation is released;
+ * (2) The time of the deallocation; and
+ * (3) The prof_tctx associated with the deallocation.
+ */
+void
+prof_recent_alloc_reset(tsd_t *tsd, extent_t *extent) {
+	malloc_mutex_lock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+	prof_recent_t *recent = extent_prof_recent_alloc_get(tsd, extent);
+	malloc_mutex_unlock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+
+	if (recent == NULL) {
+		return;
+	}
+
+	prof_tctx_t *dalloc_tctx = prof_tctx_create(tsd);
+	if (dalloc_tctx != NULL) {
+		malloc_mutex_lock(tsd_tsdn(tsd), dalloc_tctx->tdata->lock);
+		++dalloc_tctx->recent_count;
+		dalloc_tctx->prepared = false;
+		malloc_mutex_unlock(tsd_tsdn(tsd), dalloc_tctx->tdata->lock);
+	}
+
+	malloc_mutex_lock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+	recent = extent_prof_recent_alloc_get(tsd, extent);
+	if (recent != NULL) {
+		extent_prof_recent_alloc_reset(tsd, extent, recent);
+		assert(nstime_equals_zero(&recent->dalloc_time));
+		assert(recent->dalloc_tctx == NULL);
+		if (dalloc_tctx != NULL) {
+			nstime_update(&recent->dalloc_time);
+			recent->dalloc_tctx = dalloc_tctx;
+		}
+	} else if (dalloc_tctx != NULL) {
+		/* We lost the rase - the allocation record was just gone. */
+		decrement_recent_count(tsd, dalloc_tctx);
+	}
+	malloc_mutex_unlock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+}
+
+static void
+prof_recent_alloc_evict_extent(tsd_t *tsd, prof_recent_t *recent) {
+	malloc_mutex_assert_owner(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+	if (recent->alloc_extent != NULL) {
+		extent_prof_recent_alloc_reset(tsd, recent->alloc_extent,
+		    recent);
+	}
+}
+
+static void
+assert_prof_recent_alloc_non_empty(tsd_t *tsd) {
+	malloc_mutex_assert_owner(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+	assert(prof_recent_alloc_dummy != NULL);
+	assert(prof_recent_alloc_dummy->next != prof_recent_alloc_dummy);
+	assert(prof_recent_alloc_count > 0);
+}
+
+void
+prof_recent_alloc(tsd_t *tsd, const void *ptr, size_t usize,
+    extent_t *extent) {
+	assert(extent != NULL);
+	prof_tctx_t *tctx = extent_prof_tctx_get(extent);
+
+	malloc_mutex_assert_not_owner(tsd_tsdn(tsd), tctx->tdata->lock);
+	malloc_mutex_lock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+
+	prof_recent_t *reserve = NULL;
+	if (prof_recent_alloc_max == -1 ||
+	    prof_recent_alloc_count < prof_recent_alloc_max) {
+		malloc_mutex_unlock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+		reserve = (prof_recent_t *)iallocztm(tsd_tsdn(tsd),
+		    sizeof(prof_recent_t), sz_size2index(sizeof(prof_recent_t)),
+		    false, NULL, true, arena_get(tsd_tsdn(tsd), 0, false),
+		    true);
+		malloc_mutex_lock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
```

Comment:
Let's handle the OOM case right here, instead of going through the main logic.

Reply:
We can handle the OOM case here, but we still need to handle it later. In particular, if the later `reserve == NULL` check fails, it can be due to either:
- we allocated `reserve` previously but it OOMed, or
- we didn't allocate `reserve` at all previously, but after we release the lock and before we regain the lock, there has been some change so we become in need of a new node.

On the other hand, even if we OOM when we're allocating `reserve`, we might end up not needing it if things changed between the two locked period.

Thoughts?

Reply:
I see. Let's keep the current logic then. Please include some comments here saying the OOM case is handled later on.